libyara/ahocorasick.c

```diff
@@ -240,6 +381,473 @@ static int _yr_ac_state_destroy(
+      {
+        // New state with intersected input from next_state
+        memcpy(dfa_state1, input_state, sizeof(YR_AC_STATE));
+        memcpy(dfa_state1->bitmap, intersetion, sizeof(YR_BITMASK) * YR_BITMAP_SIZE);
+        _yr_ac_class_is_literal(dfa_state1);
+
+        // Updated input_state (without input from itersection)
+        memcpy(dfa_state2, input_state, sizeof(YR_AC_STATE));
+        memcpy(dfa_state2->bitmap, bitmap2, sizeof(YR_BITMASK) * YR_BITMAP_SIZE);
+        _yr_ac_class_is_literal(dfa_state2);
+        return false;
+      }
+    } // else if (input_state->type == YR_ATOM_TYPE_CLASS)
+  } // else if (next_state->type == YR_ATOM_TYPE_CLASS)
+
+  return true;
+}
+
+// In case of collision, some symbols have to be excluded from the state,
+// and the copy of the state is created
+static bool _yr_ac_exclude_from_state(
+    YR_AC_STATE* input_state,
+    YR_AC_STATE* dfa_state1,
+    YR_AC_STATE* dfa_state2,
+    YR_ARENA* arena)
+{
+  YR_AC_STATE* next1 = NULL;
+  YR_AC_STATE* next2 = NULL;
+  YR_AC_STATE* temp_state;
+  YR_AC_STATE* prev_state = NULL;
+
+  next1 = _yr_ac_next_state_typed(input_state->parent, dfa_state1->input, dfa_state1->type, dfa_state1->bitmap);
+  next2 = _yr_ac_next_state_typed(input_state->parent, dfa_state2->input, dfa_state2->type, dfa_state2->bitmap);
+
+  if ((next1 != NULL) && (next2 != NULL))
+  {
+    // Both version of the input_state already exist in the AC
+    _yr_ac_copy_path(input_state, input_state->parent, dfa_state1, arena);
+    _yr_ac_copy_path(input_state, input_state->parent, dfa_state2, arena);
+    temp_state = input_state->parent->first_child;
+
+    if (temp_state == input_state)
+    {
+      input_state->parent->first_child = temp_state->siblings;
+      _yr_ac_state_destroy_updated(input_state);
+      input_state->type = YR_ATOM_TYPE_REMOVE;
+      return true;
+    }
+
+    while (temp_state != NULL && temp_state != input_state)
+    {
+      prev_state = temp_state;
+      temp_state = temp_state->siblings;
+    }
+
+    if (prev_state != NULL)
+      prev_state->siblings = temp_state->siblings;
+
+    _yr_ac_state_destroy_updated(input_state);
+    input_state->type = YR_ATOM_TYPE_REMOVE;
+    return true;
+  }
+  else if (next2 != NULL)
+  {
+    // The updated state aleady exists in the AC
+    memcpy(input_state, dfa_state1, sizeof(YR_AC_STATE));
+    _yr_ac_copy_path(input_state, input_state->parent, dfa_state2, arena);
+  }
+  else
+  {
+    // Neither version of the input_state already exist in the AC or there is only literal state
+    memcpy(input_state, dfa_state2, sizeof(YR_AC_STATE));
+    _yr_ac_copy_path(input_state, input_state->parent, dfa_state1, arena);
+  }
+
+  return false;
+}
+
+
+// Creates deterministic AC for failure function
+static bool dfa_subtree(
+    YR_AC_STATE* state,
+    YR_AC_STATE* input_state,
+    YR_ARENA* arena)
+{
+  YR_AC_STATE* next_state;
+  YR_AC_STATE* dfa_state1 = (YR_AC_STATE*) yr_malloc(sizeof(YR_AC_STATE));
```

Comment:
When the atom is not `YR_ATOM_TYPE_CLASS` nor `YR_ATOM_TYPE_ANY` (most of the cases) these two calls to `yr_malloc` are executed unnecessarily. These two states can be allocated only when atom's type is  `YR_ATOM_TYPE_CLASS` or `YR_ATOM_TYPE_ANY`.